File: Java/1462. Course Schedule IV.java
class Solution {
    public List<Boolean> checkIfPrerequisite(int numCourses, int[][] prerequisites, int[][] queries) {
        boolean[][] reachable = new boolean[numCourses][numCourses];

        for (int[] edge: prerequisites) {
            reachable[edge[0]][edge[1]] = true;
        }

        for (int imd = 0; imd < numCourses; imd++) {
            for (int src = 0; src < numCourses; src++) {
                for (int tgt = 0; tgt < numCourses; tgt++) {
                    reachable[src][tgt] = reachable[src][tgt] ||
                        (reachable[src][imd] && reachable[imd][tgt]);
                }
            }
        }

        List<Boolean> ans = new ArrayList<>();
        for (int[] query: queries) {
            ans.add(reachable[query[0]][query[1]]);
        }

        return ans;
    }
}

File: Python3/1462. Course Schedule IV.py
class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        # reachable[i][j] means course i is a prerequisite of course j
        reachable = [[False] * numCourses for _ in range(numCourses)] 
        
        for u, v in prerequisites:
            reachable[u][v] = True

        for intermediate in range(numCourses):
            for src in range(numCourses):
                for tgt in range(numCourses):
                    reachable[src][tgt] = reachable[src][tgt] or (
                        reachable[src][intermediate] and 
                        reachable[intermediate][tgt]
                    )

        ans = []
        for u, v in queries:
            ans.append(reachable[u][v])

        return ans